docs/spec.md

```diff
@@ -116,6 +116,68 @@ types, for example, to include layouts
 ([#629](https://github.com/openxla/stablehlo/issues/629)) and sparsity
 ([#1078](https://github.com/openxla/stablehlo/issues/1078)).
 
+```ebnf
+QuantizedTensorType ::= 'tensor' '<' TensorShape QuantizedElementType '>'
+QuantizedElementType ::= '!quant.uniform' '<'
+                  QuantizationStorageType
+                  ['<' QuantizationStorageMin ':' QuantizationStorageMax '>']
+                  ':' QuantizationExpressedType
+                  [':' QuantizationDimension]
+                  ',' QuantizationParameters '>'
+QuantizationStorageType ::= IntegerType
+QuantizationStorageMin ::= IntegerConstant
+QuantizationStorageMax ::= IntegerConstant
+QuantizationExpressedType ::= FloatType
+QuantizationDimension ::= IntegerConstant
+QuantizationParameters ::= QuantizationParameter
+                         | '{' QuantizationParameter {',' QuantizationParameter} '}'
+QuantizationParameter ::= QuantizationScale ':' QuantizationZeroPoint
+QuantizationScale ::= FloatConstant
+QuantizationZeroPoint ::=  IntegerConstant
+```
+
+**Quantized element types** represent integer values of a **storage type** in
+the range from `storage_min` to `storage_max` (inclusive) that correspond to
+floating-point values of an **expressed type**. For a given integer value `i`,
+the corresponding floating-point value `f` can be computed as
+`f = (i - zero_point) * scale`, where `scale` and `zero_point` are called
+**quantization parameters**. The `storage_min` and `storage_max` are optional
+in the grammar, but have default values of `min_value(storage_type)` and
+`max_value(storage_type)` respectively. Quantized element types have the
+following constraints:
```

Comment:
>> What is the semantics of storage_min and storage_max?

The semantics is to make sure that the quantized values are within this range.  This semantics is enforced by the individual ops. 

>> Are operators which produce quantized values supposed to clamp them to [storage_min, storage_max]?

Yes.

>> Are StableHLO consumers supposed to validate that quantized tensors that are inputs to the model don't have values outside of [storage_min, storage_max]?

The spec in its current form, uses the individual ops to apply the semantics of the `[storage_min, storage_max]`, which is to clamp the quantized values. There is no constraint in the input data and IMO the consumers can still be compliant to the spec even if they do not have the validation.

Reply:
For some cases, we also use the [-7, 8] range with eight bit integer storage to mimick four bit quantization or other bit scheme quantization with wider integer storage.

For popular eight bit quantization cases, as you said, the typical ranges, [-128, 128] [-127, 128], are popular. However, IMO, for various reasons (e.g, special quantization treatments, research purpose) it might be better not to limit allowed ranges with one or two ranges only in the language specification.

Reply:
This is a good point! Let's discuss more in #1406. In the meanwhile, we removed the spec for quantized AddOp to sidestep having an opinion on this in this specific PR.